### PR TITLE
Add an example for Magnum Audio

### PR DESCRIPTION
**Pull request by Squareys**
Hi @mosra !

Just to let you know I am currently working on this, here's an early PR for the magnum-audio example possibly coming in near future (in case it's good enough to pass review ;P ).

This is the first steps for getting the following line added to magnums description: "Audio is not an ugly stepchild either: [...]", we'll see if I can make this happen. I am hoping to implement OpenAL extension support in the future, but I don't know if I'll do more than the HRTF extension.

I hope you're as exited about this as I am ;)

Greetings, Squareys

### CODE REVIEW

**Comment by mosra** [comment]
Wow, didn't expect this :)

Yeah, audio currently *is* an ugly stepchild, the original implementation was done in a hurry for some game jam and I hadn't touched it since. I actually have some work-in-progress example in my local copy, but never managed to finish it, mainly because I discovered that the `WavAudioImporter` is not able to parse real-world files :/

What I planned but didn't have a time to do:
 - adding `Playable` scene graph feature to `Magnum::Audio` namespace that would manage `Audio::Source` and properly update its position, orientation (velocity?) on transformation change, so the user doesn't have to implement this every time from scratch
 - adding `AudioSourceRenderer` or something like that to `Magnum::DebugTools` namespace, which would visualize position, orientation, cone angle, gain etc. using some simple line art

**Comment by Squareys** [comment]
@mosra I am currently trying to use `Drawable2D`, `Flat2D`, `MatrixTransformation2D` and such. I cannot seem to get anything rendered to the screen (grey screen of death? :D ). I must be missing something.

> because I discovered that the WavAudioImporter is not able to parse real-world files :/

That doesn't sound too good. I will see what I can do if I discover that aswell.

> adding Playable scene graph feature to Magnum::Audio namespace

Yes, and something which manages the listener position.

> Adding AudioSourceRenderer or something like that to Magnum::DebugTools namespace

That is a very good idea. Obviously "hearing" if the sound is coming from the right direction wont help too much when debugging :dancer: 

> my example would then be something similar to the bullet one, just displaying the audio sources using the debug renderer, playing them and changing their position based on user input.

After blowing up the example with image loading and such, I understand what you mean. Since this depends on the future changes/additions to the Audio module, I will implement this as originally planned first, and adapt it to debug renderers once the job in Magnum it self has been done. Is this okay with you?

**Comment by mosra** [comment]
WavAudioImporter is not able to parse real-world files becuase I incorrectly assumed that [there aren't any additional RIFF chunks](https://github.com/mosra/magnum/blob/master/src/MagnumPlugins/WavAudioImporter/WavImporter.cpp#L61) -- the code should just skip the unknown ones and read the data. I don't remember the particulars, sorry, long time since I last looked at it, but can certainly help you if you have any questions.

Oh, there's also [stb_vorbis](https://github.com/nothings/stb/blob/master/stb_vorbis.c) which might be easy enough to convert into a plugin for opening OGG files (as compared to MP3/Lame) ... or just used directly in the example if you want to have something quickly up and running :)

> Since this depends on the future changes/additions to the Audio module, I will implement this as originally planned first, and adapt it to debug renderers once the job in Magnum it self has been done. Is this okay with you?

Sure! :)

**Comment by Squareys** [comment]
@mosra That was *alot* of fun! I just need to update the readme and then this is ready to merge!

You may want to listen into the wav (which is loaded perfectly fine btw). I got it from freesound.org: http://freesound.org/people/dkmedic/sounds/104457/
In case you find it inappropriate, I will use a different one. If you have a preference, just send me a link ;)

File: src/audio/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+    Shaders
+    Sdl2Application)
+
+set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} ${CORRADE_CXX_FLAGS}")
+include_directories(${MAGNUM_INCLUDE_DIRS} ${MAGNUM_APPLICATION_INCLUDE_DIRS} ${MAGNUM_AUDIO_INCLUDE_DIRS} ${CMAKE_CURRENT_BINARY_DIR})
+
+configure_file(${CMAKE_CURRENT_SOURCE_DIR}/configure.h.cmake
+               ${CMAKE_CURRENT_BINARY_DIR}/configure.h)
+
+corrade_add_resource(TexturedTriangleData resources.conf)
```
Comment (mosra) [inline src/audio/CMakeLists.txt]: Probably a copypaste bug, if you enable also the `textured-triangle` example, CMake then complains about two targets with the same name (should be `AudioData`?)

Comment (Squareys) [inline src/audio/CMakeLists.txt]: Hah, great :P Yeah, indeed.

File: src/audio/AudioExample.cpp [hunk 2]
```diff
@@ -0,0 +1,289 @@
...
+    PluginManager::Manager<Trade::AbstractImporter> imageManager{MAGNUM_PLUGINS_IMPORTER_DIR};
+    PluginManager::Manager<Audio::AbstractImporter> audioManager{MAGNUM_PLUGINS_AUDIOIMPORTER_DIR};
+
+    if(!(imageManager.load("TgaImporter") & PluginManager::LoadState::Loaded))
+        std::exit(1);
+    std::unique_ptr<Trade::AbstractImporter> importer = imageManager.instance("TgaImporter");
+
+    if(!(audioManager.load("WavAudioImporter") & PluginManager::LoadState::Loaded))
+        std::exit(1);
+    std::unique_ptr<Audio::AbstractImporter> wavImporter = audioManager.instance("WavAudioImporter");
```
Comment (mosra) [inline src/audio/AudioExample.cpp]: You might want to use [PluginManager::Manager::loadAndInstantiate()](http://mosra.cz/blog/corrade-doc/classCorrade_1_1PluginManager_1_1Manager.html#a1b831c65abd0581f9d336bd3ad820058), it's less repetitive if you don't need to do extensive checking.

File: src/audio/README.md [hunk 3]
```diff
@@ -0,0 +1,12 @@
...
+The sound source can be moved around the listener by clicking on the screen.
+
+![Audio](audio.png)
+
+Key shortcuts
+-------------
+
+**Left-click**: Set sound source position.
+
+
```
Comment (mosra) [inline src/audio/README.md]: Trailing newlines?

**Comment by Squareys** [comment]
@mosra I declare this PR as ready from my side.

File: src/audio/AudioExample.cpp [hunk 4]
```diff
@@ -0,0 +1,289 @@
...
+    _sourceTop.reset(new TexturedDrawable2D(_mesh, _shader, _textureSource, _sourceTopObject, _drawablesTop));
+
+    /* We do not need depth test, drawing order is fine. */
+    Renderer::disable(Renderer::Feature::DepthTest);
+
+    /* Enable blending for alpha channel in textures. */
+    Renderer::enable(Renderer::Feature::Blending);
+    Renderer::setBlendFunction(Renderer::BlendFunction::SourceAlpha, Renderer::BlendFunction::OneMinusSourceAlpha);
+
+    Renderer::setClearColor({0.6, 0.6, 0.6, 1.0});
```
Comment (mosra) [inline src/audio/AudioExample.cpp]: I would maybe prefer that all examples have the same (=default) background color, to make them look similarly on the [showcase page](http://mosra.cz/blog/magnum-showcase.php).

**Comment by mosra** [comment]
Added a few comments inline, other than that I'm more than happy. Great job again!

> I hope it's okay that I added the audio example here myself.

Sure, no problem :)

File: src/audio/AudioExample.cpp (hunk 4 above)
Comment (Squareys) [inline src/audio/AudioExample.cpp]: Alright, I can understand that.

File: src/audio/AudioExample.cpp (hunk 2 above)
Comment (Squareys) [inline src/audio/AudioExample.cpp]: You may want to fix that in TexturedTriangle example too, maybe. I think I used that for reference.

**Comment by Squareys** [comment]
@mosra All your inline comments have been respected.

File: src/audio/AudioExample.cpp (hunk 2 above)
Comment (mosra) [inline src/audio/AudioExample.cpp]: ...and everywhere else except for Viewer, which somehow has it right :) 

Good point, thank you, noted.

**Comment by Squareys** [comment]
Thanks! :)

**Comment by mosra** [comment]
Me too! :)

**Pull request merged**